server/src/main/java/org/opensearch/index/engine/EngineConfigFactory.java

```diff
@@ -124,7 +212,9 @@ public EngineConfig newEngineConfig(
             mergePolicy,
             analyzer,
             similarity,
-            this.codecService != null ? this.codecService : codecService,
+            this.codecServiceFactory != null
+                ? this.codecServiceFactory.createCodecService(new CodecServiceConfig(indexSettings, mapperService, logger))
```

Comment:
I think we should only build CodecService in one place and always use the factory. That way, we can avoid ternary statements like the above. For default CodecServiceFactory, we can set it like I did [here](https://github.com/jmazanec15/OpenSearch/commit/165f50a3bf9fa067f777acd85770608299362534#diff-5eecdf54afeae66f50a535c5d634cf34bc5811f67639189dbc13f3b65e7573d4R119).

Im somewhat conflicted on where to build CodecService. I think there are 2 options: 

#### 1. Call `this.codecServiceFactory.createCodecService` here and dont take in CodecService as arg to newEngineConfig

The pro's to this approach are that we can hide details about CodecService from all other parts of code - in IndexShard we dont need to do anything with codec service. Its all handled by EngineConfigFactory. 

The cons are that a new CodecService would get created every time this method gets called. I'm not really sure if there will be side effects to this. In IndexShard, this gets called in `resetEngineToGlobalCheckpoint` and `innerOpenEngineAndTranslog`. One potential issue could arise if a user changes IndexSettings between these 2 operations and that impacts how the CodecServiceFactory builds the CodecService (maybe this is a feature not a bug?).

#### 2. Call it in IndexShard's constructor by exposing getter on EngineConfigFactory's codecServiceFactory

Alternatively, we could create the codec service in IndexShards constructor. This is how it is being done now for the default. To do this, we would need to expose a getter method for EngineConfigFactory to get its codecServiceFactory and then call `this.codecServiceFactory.createCodecService()`. Then, we wouldnt change anything in `EngineConfigFactory.newEngineConfig()`. The getter, however, feels a little bit out of place here. I took this approach [here](https://github.com/jmazanec15/OpenSearch/commit/165f50a3bf9fa067f777acd85770608299362534#diff-acf4481a07fc77c1b5d6eaca20a6a1ce1b171615c4349b67bd23b2142a2f0ccbR330). 

I think I am leaning towards approach 2 as opposed to 1, to keep consistent with how its being done now.

Reply:
@jmazanec15 I think I got your idea, the challenge here is to preserve the semantics of the `CodecService` instantiation, I just pushed the solution which I believe is the best effort: 
 - new method `EngineConfigFactory::newCodecServiceOrDefault` which basically either create a new `CodecService` instance from factory or uses the existing one
 - `IndexShard::newEngineConfig` uses  `EngineConfigFactory::newCodecServiceOrDefault` to pass the `CodecService` instance to `EngineConfigFactory::newEngineConfig`
 - `EngineConfigFactory::newEngineConfig` has the fallback in case passed `CodecService` is `null` but there is `CodecServiceFactory` configured - the old way to create `CodecService` without `MapperService` and `Logger`

The APIs stay unchanged, thank you

Reply:
That makes sense thanks!